client/task_runner.go

```diff
@@ -401,7 +402,7 @@ func (r *TaskRunner) run() {
 				close(stopCollection)
 
 				// Store that the task has been destroyed and any associated error.
-				r.setState(structs.TaskStateDead, structs.NewTaskEvent(structs.TaskKilled).SetKillError(err))
+				r.setState(structs.TaskStateDead, r.destroyEvent.SetKillError(err))
```

Comment:
Revert this back to what it was and then at the top of the `case <-r.destroyCh` block add a check:

```
// If we have a destroy event it is not due to the normal life-cycle of the task. Set the reason and continue with 
// the normal kill behavior.
if r.destroyEvent != nil {
   r.SetState(structs.TaskStateRunning, r.destroyEvent)
}
```